File: app/src/main/java/com/example/solar_alarm/Data/SolarAlarmDatabase.java
package com.example.solar_alarm.Data;

import static com.example.solar_alarm.Data.Migrations.StaticDataMigration.MIGRATION_1_2;

import android.content.Context;
import android.os.Build;

import androidx.annotation.RequiresApi;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.TypeConverters;

import com.example.solar_alarm.Data.Daos.LocationDao;
import com.example.solar_alarm.Data.Daos.SolarAlarmDao;
import com.example.solar_alarm.Data.Daos.SolarTimeDao;
import com.example.solar_alarm.Data.Daos.TimezoneDao;
import com.example.solar_alarm.Data.Tables.AlarmType;
import com.example.solar_alarm.Data.Tables.Location;
import com.example.solar_alarm.Data.Tables.SolarAlarm;
import com.example.solar_alarm.Data.Tables.SolarTime;
import com.example.solar_alarm.Data.Tables.SolarTimeType;
import com.example.solar_alarm.Data.Tables.TimeUnitType;
import com.example.solar_alarm.Data.Tables.Timezone;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@RequiresApi(api = Build.VERSION_CODES.O)
@Database
(
    entities =
    {
        AlarmType.class,
        Location.class,
        SolarAlarm.class,
        SolarTime.class,
        SolarTimeType.class,
        TimeUnitType.class,
        Timezone.class
    },
    version = 1,
    exportSchema = false
)

@TypeConverters({Converters.class})
public abstract class SolarAlarmDatabase extends RoomDatabase
{
    public abstract LocationDao locationDao();
    public abstract SolarAlarmDao solarAlarmDao();
    public abstract SolarTimeDao solarTimeDao();
    public abstract TimezoneDao timezoneDao();

    private static volatile SolarAlarmDatabase INSTANCE;
    private static final int NUMBER_OF_THREADS = 4;
    public static final ExecutorService databaseWriteExecutor = Executors.newFixedThreadPool(NUMBER_OF_THREADS);

    public static SolarAlarmDatabase getDatabase(final Context context)
    {
        if (INSTANCE == null)
        {
            synchronized (SolarAlarmDatabase.class)
            {
                if (INSTANCE == null)
                {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(), SolarAlarmDatabase.class, "SolarAlarmDatabase")
                            .addMigrations(MIGRATION_1_2).build();
                }
            }
        }

        return INSTANCE;
    }
}

File: app/src/main/java/com/example/solar_alarm/Data/SolarAlarmDatabase.kt
package com.example.solar_alarm.Data

import androidx.annotation.RequiresApi
import android.os.Build
import com.example.solar_alarm.Data.Daos.LocationDao
import com.example.solar_alarm.Data.Daos.StaticDataDao
import com.example.solar_alarm.Data.Daos.SolarAlarmDao
import com.example.solar_alarm.Data.Daos.SolarTimeDao
import androidx.room.Database
import androidx.room.RoomDatabase
//import com.example.solar_alarm.Data.Migrations.StaticDataMigration
import androidx.room.TypeConverters
import kotlin.jvm.Volatile
import androidx.room.Room
import android.content.Context
import androidx.sqlite.db.SupportSQLiteDatabase
import com.example.solar_alarm.Data.Tables.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

@RequiresApi(api = Build.VERSION_CODES.O)
@Database(entities = [OffsetType::class, Location::class, SolarAlarm::class, SolarTime::class, SolarTimeType::class], version = 1, exportSchema = false)
@TypeConverters(Converters::class)
abstract class SolarAlarmDatabase : RoomDatabase()
{
    abstract fun locationDao(): LocationDao
    abstract fun solarAlarmDao(): SolarAlarmDao
    abstract fun solarTimeDao(): SolarTimeDao
    abstract fun alarmDisplayDataDao(): AlarmDisplayDataDao
    abstract fun staticDataDao(): StaticDataDao

    companion object
    {
        @Volatile
        private var INSTANCE: SolarAlarmDatabase? = null
        private const val NUMBER_OF_THREADS = 4
        val databaseWriteExecutor: ExecutorService = Executors.newFixedThreadPool(NUMBER_OF_THREADS)

        fun getDatabase(context: Context, scope: CoroutineScope): SolarAlarmDatabase
        {
            // if the INSTANCE is not null, then return it,
            // if it is, then create the database
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    SolarAlarmDatabase::class.java,
                    "word_database"
                )
                    // Wipes and rebuilds instead of migrating if no Migration object.
                    // Migration is not part of this codelab.
                    .fallbackToDestructiveMigration()
                    //.addMigrations(StaticDataMigration.Companion.MIGRATION_1_2)
                    //.addCallback(SolarAlarmDatabaseCallback(scope))
                    .build()
                INSTANCE = instance
                // return instance
                instance
            }
        }
    }

    private class SolarAlarmDatabaseCallback(private val scope: CoroutineScope) : RoomDatabase.Callback()
    {
        override fun onCreate(db: SupportSQLiteDatabase)
        {
            super.onCreate(db)
            INSTANCE?.let { database -> scope.launch { populateDatabase(database.locationDao()) } }
        }

        suspend fun populateDatabase(locationDao: LocationDao) {
            // Delete all content here.
            //locationDao.deleteAll()

            // Add sample words.
            //var word = Word("Hello")
            //wordDao.insert(word)
            //word = Word("World!")
            //wordDao.insert(word)

            // TODO: Add your own words!
        }
    }
}
